src/controller/UserGeo.ts

```diff
@@ -0,0 +1,142 @@
+const UserGeoSchema = Joi.object({
+  id: UserSchemaId,
+  longitude: Joi.number().min(-180).max(180).required(),
+  latitude: Joi.number().min(-85.05112878).max(85.05112878).required(),
+});
+
+// -*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
+// APIS
+// -*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
+/**
+ * @swagger
+ * /usergeo:
+ *   post:
+ *     tags:
+ *       - UserGeo
+ *     description: save user geo data, if data not exists it would be created, if exists then updated
+ *     operationId: saveUserGeo
+ *     consumes:
+ *       - application/json
+ *     produces:
+ *       - application/json
+ *     parameters:
+ *       - name: data
+ *         description: user geo data
+ *         in: body
+ *         required: true
+ *         schema:
+ *           $ref: '#/definitions/UserGeo'
+ *     responses:
+ *       200:
+ *         description: request processed without error
+ *         schema:
+ *           $ref: '#/definitions/OkResponse'
+ */
+export const saveUserGeo = async (ctx: Koa.Context) => {
+  const body = (ctx.request as any).body as UserGeoModel;
+  const {error} = await validateWithJoi(UserGeoSchema, body);
+  if (error) {
+    return responseError("UserGeoController", "saveUserGeo", ctx, -1, error);
+  }
+
+  try {
+    await UserGeoDao.saveUserGeo(body);
+
+    return responseNormal("UserGeoController", "saveUserGeo", ctx, "OK");
+  } catch (err) {
+    return responseError("UserGeoController", "saveUserGeo", ctx, -1, err);
+  }
+};
+
+/**
+ * @swagger
+ * /usergeo/nearby/{id}:
+ *   get:
+ *     tags:
+ *       - UserGeo
+ *     description: get nearby user ids by target userId
+ *     operationId: getUser
+ *     consumes:
+ *       - application/json
+ *     produces:
+ *       - application/json
+ *     parameters:
+ *       - name: id
+ *         description: user id
+ *         in: path
+ *         required: true
+ *         type: string
+ *         example: "29256e52-9ca2-49d5-8b4e-f42578ccae68"
+ *     responses:
+ *       200:
+ *         description: user data
+ *         schema:
+ *           $ref: '#/definitions/NearbyResponse'
+ */
+export const getNearbyUserIds = async (ctx: Koa.Context) => {
+  const {error} = await validateWithJoi(UserSchemaId, ctx.params.id);
+  if (error) {
+    return responseError("UserGeoController", "getNearbyUserIds", ctx, -1, error);
+  }
+
+  try {
+    const result = [] as string[];
+    const followers = await UserLinkDao.getFollowers(ctx.params.id);
+    const nearbyIds = await UserGeoDao.getNearbyUserIds(ctx.params.id);
+
+    for (const [, id] of nearbyIds.entries()) {
+      if (id !== ctx.params.id // userId self would be returned by redis, so shall be filtered here
+        && followers.indexOf(id) !== -1) { // only user's "friends" (follower, since it's passive action) shall be displayed
+        result.push(id);
+      }
+    }
```

Comment:
Let's assume here we need to support millions of Users with geolocation info and now the response of `getNearbyUserIds` is not fast enough, can you figure out a way to optimize it?

Reply:
We could save user position into `block`.

Geohash has precision, low precision means a `block`:

![block1](https://user-images.githubusercontent.com/195293/95644188-83ce7180-0ae7-11eb-9586-117877c43dc4.png)

Use this low precision geohash value as key. Save user data into corresponding `block`.

When we need to retrieve nearby users. Use position of user to calculate which `block` user belongs to. Retrieve data in this block and calculate nearby users via geohash values. If the radius is bigger than one `block`, then also retrieving data of surrounding `block`s.

![block2](https://user-images.githubusercontent.com/195293/95644273-1e2eb500-0ae8-11eb-8693-7054db45f5ba.png)

Reply:
Thanks, I see your idea, can you also give a time complexity with [big O notation](https://en.wikipedia.org/wiki/Time_complexity) and explanation correspondingly with the new solution?

Reply:
Case 1: update user position:

1. remove old position: `ZREM` `O(log(N))`, N being the number of elements in the sorted set
2. add new position: `GEOADD` `O(log(N))`, N being the number of elements in the sorted set

Total complexity: `O(log(N))`, N being the number of elements in the sorted set

---

Case 2: get nearby users, and searching radius is inside a block

1. get nearby members inside the sorted set: `GEORADIUSBYMEMBER` `O(N+log(M))`, N is the number of elements inside the bounding box of the circular area delimited by center and radius and M is the number of items inside the sorted set
2. get user friends list: `SMEMBERS` `O(N)`, N is the set cardinality
3. compare fetched nearby users to friends list, to determine nearby friends: worst `O(N*M)` (as codes below), N is nearby users count fetched and M is the length of friends list

```typescript
for (const [, nearbyUserId] of nearbyUsers.entries()) {
	if (friendIds.indexOf(nearbyUserId) !== -1) {
		// ...
	}
}
```

Total complexity: `O((N+log(M))*K)`, N is the number of elements inside the bounding box of the circular area delimited by center and radius and M is the number of items inside the sorted set and K is the length of friends list

---

Case 3: get nearby users, and radius is bigger than a block(meanings shall calculate 9 blocks)

1. retrieve all members(userId) from 9 blocks(sorted set): `ZRANGE` `O(log(N)+M)`, with N being the number of elements in the sorted set and M the number of elements returned
2. retireve all position data of members from 9 blocks: `GEOPOS` `O(log(N))`, where N is the number of elements in the sorted set
	* since GEOPOS requires all member values as params, so step 1 is required
3. calculate the distence between members and target user to filter "nearby users": `some geohash lib function` `O(N)`, where N is the number of all memebers in 9 blocks
4. same as step 2 of previous case
5. same as step 3 of previous case

Total complexity: `O((log(N)+M)*K)`, N being the number of elements in the sorted set(9 blocks) and M the number of elements returned(9 blocks) and K is the length of friends list